### PR TITLE
Fix redeploying SageMaker Model Monitoring module.

### PR DESCRIPTION
**Pull request by clokep**
## Describe your changes

Fixes redeploying SageMaker Model Monitoring. Previously if you modified a parameter then CloudFormation would error due to attempting to replace resources (via a create, then delete) using the same name. This adds a timestamp to names so that they'll be unique. Ideally we would allow CloudFormation to generate the names, but this does not seem to work -- CloudFormation is not properly waiting for the job definition to be created before attempting to access the name / propagating it to the monitoring schedule.

The downside of this approach is the monitoring stack will essentially be rebuilt for every update, even if no changes are made. This is double bad since updates will fail if any of the monitoring jobs are running. A potential solution to this could be to use a hash of the input parameters instead of a timestamp.

## Checklist before requesting a review

- [ ] I updated `CHANGELOG.MD` with a description of my changes - *not updated since the SageMaker Monitoring module hasn't yet been in a release*
- [x] If the change was to a module, I ran the code validation script (`scripts/validate.sh`)
- [x] If the change was to a module, I have added thorough tests
- [x] If the change was to a module, I have added/updated the module's README.md
- [x] If a module was added, I added a reference to the module to the repository's README.md
- [x] I verified that my code deploys successfully using `seedfarmer apply`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: AIOpsModuleTestsInfrastruct-doMLXEYsnmxr
* Commit ID: 22fdfdb11ffedea46e8942320e58964c1b9ef31c
* Result: SUCCEEDED
* [Build Logs](https://ythfngxmh6.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=443fc959-b0e0-4965-8f52-8d436e49779f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: AIOpsModuleTestsInfrastruct-doMLXEYsnmxr
* Commit ID: 5d792a23bc8fad5c3909e49a9ab56c6f545a926a
* Result: SUCCEEDED
* [Build Logs](https://ythfngxmh6.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=a6360992-b2e2-44de-8905-d2171f92e091%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by clokep** [comment]
> The downside of this approach is the monitoring stack will essentially be rebuilt for every update, even if no changes are made. This is double bad since updates will fail if any of the monitoring jobs are running. A potential solution to this could be to use a hash of the input parameters instead of a timestamp.

This was easy enough to fix by using a hash of the input parameters. (Note that the hash doesn't need to be secure due to not being used in a security context.) There's a risk of a collision, especially since the hash gets truncated due to the 63 character limit on resource names. Unsure if this is better or not. Seems like the options are:

1. Always redeploy due to generating a new job definition name each time seedfarmer is run. This risks a failed deploy it the monitoring jobs are running during a deploy.
2. Deploy only when input parameters change. Risks a failed deploy if there's a hash collision based on the input parameters resulting in the same hash.

Is there another option? Is there an opinion about which option is better? 1 is safer, 2 is will result in less churn.

**Comment by kukushking** [comment]
Thanks @clokep, I think option 2 works. I would not sweat too much about the risk in approach 1 if we instruct the user in the README to stop the schedule if there is a pending deployment, but I agree option 2 is less churn. We can always review and update this based on user feedback.

**Review by kukushking — APPROVED** [review]

File: modules/sagemaker/sagemaker-model-monitoring/sagemaker_model_monitoring/utils.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+    Any non-string arguments are cast to a string.
+    """
+    hash = md5()
+    for arg in args:
+        if not isinstance(arg, str):
+            arg = str(arg)
+        hash.update(arg.encode())
+        hash.update(b"\x00")
+    return hash.hexdigest()[:10]
```
Comment (kukushking) [inline modules/sagemaker/sagemaker-model-monitoring/sagemaker_model_monitoring/utils.py:17]: nit - I think you can just cast all to string and would be able to re-write this into a one-liner list comprehension

**Comment by clokep** [comment]
Oh good call that we should document either way the failure about it running. I think it could still fail even with option 2, will need to try to do that on purpose to confirm and then document.

**Review by kukushking — APPROVED** [review]
Please update CHANGELOG, README & ship it :)

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: AIOpsModuleTestsInfrastruct-doMLXEYsnmxr
* Commit ID: 9f31e567a2528960887c9c3e14c079074fe0cce8
* Result: SUCCEEDED
* [Build Logs](https://ythfngxmh6.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=5b0a8a3a-e2a9-40ea-bed4-3a97892d46e2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by clokep** [comment]
> Please update CHANGELOG, README & ship it :)

We chatted a bit about this offline and will leave this out of the changelog since the unreleased changes already include that this module was just *added*.

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: AIOpsModuleTestsInfrastruct-doMLXEYsnmxr
* Commit ID: 19428e7fb546ddb0c81a9317c3654a02b4c95959
* Result: SUCCEEDED
* [Build Logs](https://ythfngxmh6.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=9ef4aca5-3fef-435e-a78d-3945a948c274%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by kukushking — APPROVED** [review]
Thank you!

**Pull request merged**